enhancements/image-references.md

```diff
@@ -0,0 +1,228 @@
+
+The only acceptable process for creating an operator bundle will be to run the
+`operator-sdk bundle create` CLI, providing the manifests, metadata and image
+sources as input to the creation process.
+
+### Proposed UX
+
+Operator manifest authors write a manifest that refers to images using some
+opaque string [ex](https://github.com/openshift/machine-config-operator/blob/master/install/0000_80_machine-config-operator_04_deployment.yaml#L22),
+and provide an `image-references` file alongside their manifests that declares
+which strings inside of their manifest are referring to pull specifications of
+images and names each occurence [ex](https://github.com/openshift/machine-config-operator/blob/98d9ba6841eb4811ed6f4d4de7016ea83c131c54/install/image-references#L7-L10).
+
+The `metadata/image-references` file holds data in the following format, mapping
+common names of container imags to their string placeholders in the manifests:
+
+```yaml
+imageReferences:
+- name: common-name
+  placeholder: registry.svc.ci.openshift.org/openshift:image
+```
+
+This mapping, therefore, defines what needs to be replced in manifests when
+they are bundled and identifies each replacement with a name. When
+`operator-sdk bundle create` runs, it will require as input a second mapping
+from those names to literal image pull specifications and will run the replacement.
+In this manner, the configuration provided by the manifest author remains static
+regardless of the eventual replacement that a build system will execute.
+
+The second mapping required at build-time, known as `image-replacements`, will take
+a similar form, mapping the common names of images to their explicitly resolved pull
+specifications. This file will have the following format:
+
+```yaml
+imagePullSpecs:
+- name: common-name
+  pullSpec: quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:298d0b496f67a0ec97a37c6e930d76db9ae69ee6838570b0cd0c688f07f53780
+```
+
+`operator-sdk bundle create` will use this chain of mapping to perform replacements
+in the manifests before creating a bundle image layer. The layer creation will also
+be shared logic with `oc adm release new` in order to allow both processes to build
+image layers without requiring the use of a containe runtime, other build system, any
+elevated permissions, privleges, capacities or SELinux roles. As the output image
+layer in both cases is `FROM scratch` and simply contains manifest data, this build
+process is simple and producing the layer by creating the underlying tar bundle does
+not come with risks.
+
+The bundle creation process will furthermore create as output an `images` metadata file
+that will expose the build-time inputs used to create the bundle, so that downstream
+consumers can access this data as necessary. This file will also allow the build system
+to add any further metadata necessary. The format for the `images` file is a strict
+super-set of the `image-replacements` file:
```

Comment:
Why can't `image-replacements` be used directly here? What "metadata" do you envision being required that isn't stored elsewhere.

Reply:
@smarterclayton @wking @masuzlik could expand on how this is used for OpenShift release payloads.

Reply:
maszulik is @soltysh

Reply:
Do we have this for OpenShift release images?  We have [a separate `release-metadata` file in those][1], maybe you're just making space for additional metadata like that in a single file?

The most important metadata we store in `release-metadata` is the release image's version string (e.g. `4.1.0`).  Having that inside the signed release image means that consumers can say "would you like to update to 4.1.0?" without having to worry about whether the by-digest pullspec they'll use is actually for some different release.  But if you always bundle into container images you could theoretically store that metadata as [labels][2] on the image too.

[1]: https://github.com/openshift/cluster-version-operator/blob/62ff8e3e51cd4bfe6cd3b8cb6400a28ada1b183d/docs/user/reconciliation.md#release-image-content
[2]: https://github.com/opencontainers/image-spec/blame/v1.0.1/config.md#L159-L162

Reply:
Ah, I was mistaken in my understanding, then. I'll update to use a `release-metadata` file.